app/handlers/personal/quick_self_test.py

```diff
@@ -0,0 +1,264 @@
+                'context_2': UUID('7e831026-0c60-448e-9669-3c085d5f6903') *
+            }
+            * - The value is unchanged throughout the cycle
+
+
+    Returns:
+        msg:
+            see base class
+            Source: https://core.telegram.org/bots/api#message
+    """
+
+    if state_data["end_time"] < datetime.now(tz=ZoneInfo("UTC")):
+        await state.clear()
+        return await msg.answer(text="Training time has expired.")
+
+    card: dict[str, Any] = await get_actual_card(state_data["user_id"])
+    if not card:
+        await state.clear()
+        return await msg.answer(text="Run out of words to study")
+
+    texts: list[str] = [card["item_1"], card["item_2"]]
+    random.shuffle(texts)
+    text_for_show, correct_answer = texts
+
+    hidden_answer: str = HtmlDecoration().spoiler(correct_answer)
+    answer: str = text_for_show + "\n" + hidden_answer
+    await state.set_state(FSMSelfTest.selftest)
+    return await msg.answer(
+        text=answer,
+        parse_mode="HTML",
+        reply_markup=know_dont_know(
+            card_id=card["id"],
+            memorization_stage=card["memorization_stage"],
+            repetition_level=card["repetition_level"],
+        ),
+    )
+
+
+async def handler_know_dont_know(
+        callback_query: types.CallbackQuery,
+        callback_data: KnowDontKnowCallbackData,
+        state: FSMContext,
+) -> Union[types.Message, bool]:
+    """Processes the result for self-test mode keyboard work.
+
+    Up to dates DB data in Card table according to:
+    - memorization_stage
+    - repetition_level
+    - result answer (True, False).
+    If the user did not answer in the current session, the data
+    in the DB is not updated.
+
+    Parameters:
+        callback_query:
+            see base class
+        callback_data:
+            see base class
+        state:
+            State Machine where save data current session:
+            {
+                'end_time': datetime.datetime(2023, 1, 30, 9, 45, 22, 55017, tzinfo=zoneinfo.ZoneInfo(key='UTC')),
+                'user_id': UUID('08ef015a-36d8-49e7-8b0b-8d86f951a78e'), *
+                'texts_context_1': [{'text': 'дієта'}, {'text': 'сюрприз'}, ...] *
+                'texts_context_2': [{'text': 'pen'}, {'text': 'flat'}, ...] *
+                'context_1': UUID('86edaed5-336b-4b3b-b103-7dbb36f1ad34'), *
+                'context_2': UUID('7e831026-0c60-448e-9669-3c085d5f6903'), *
+                'correct_translation': {'text_for_show': 'виправити', 'correct_answer': 'fix'} **
+            }
+            * - The value is unchanged throughout the cycle
+            ** - Value change on every cycle
+    Return:
+        Updates the previous answer, removes the keyboard from it
+        and adds the correct word and symbol to the answer text -
+        the answer was correct or not.
+    """
+    if callback_query.message is None:
+        return await callback_query.answer("Pay attention the message is too old.")
+
```

Comment:
```suggestion
    state_data: dict[str, Any] = await state.get_data()
    correct_translation: Optional[dict] = state_data.get("correct_translation")

    # when the user replied not in the current session
    if (
        correct_translation is None
        or correct_translation["text_for_show"] != callback_query.message.text
    ):
        await callback_query.answer("Message is outdated.")
```
Пропоную використати цей код і тут, щоб при натисках на клавіатуру які були в минулих сесіях не вносилися зміни в БД (поки ми так вирішили). Певно реалізацію треба підкорегувати, але сама ідея залишиться.

Reply:
Тобто це if для того випадку, коли в state_date  пусто, або юзер зараз має перекладити інше слово, але йому щось торкнуло тикнути на переклад "старого" слова ?
Може тоді там і кінці return  поставити, щоб не запускалось  set_res_studying_card(
            Card(...
            ?
            Чи я не правильно зрозуміла ціль цього нового коду?

Reply:
agree.
done

Reply:
зробила